### PR TITLE
fix: add route for underlay subnet which enables u2o and disables LB

### LINKED ISSUE #4038: [BUG] 关闭Loadbalancer并开启U2O后无法访问Service
### Kube-OVN Version

v1.13.0

### Kubernetes Version

v1.25.3

### Operation-system/Kernel Version

"Ubuntu 22.04.4 LTS" 6.5.0-35-generic

### Description

前提：关闭LoadBalancer
1、目前的设计中，只支持开启U2O连接ovn-cluster的Underlay子网访问Overlay网络，其中Underlay子网开启U2O后会通过lrp和ovn-cluster互联，lrp上分配IP地址，开启该功能后创建的Pod/虚机网关会指向lrp的IP，因此只能支持开启U2O后创建的Pod/虚机访问Overlay网络。
2、开启U2O后Underlay访问Overlay是通过OVS进行转发的，而为了实现Underlay访问K8S的ClusterIP类型的Service（Service的endpoints是Overlay网络中的一些Pod），会在ovn-cluster添加策略路由，将每个node上的Underlay访问ClusterIP的流量路由至node上的ovn0上。

以下给出请求流量的Tcpdump相关路径：
b2ba4d537f3a_h <--> underlay的pod网卡的veth pair（2.2.98.0/24）
7fc526e9ec01_h <--> ovn-default的pod网卡的veth pair（10.222.X.X/18）
Cluster IP range (10.233.X.X/16)

b2ba4d537f3a_h P       2.2.98.7.42462 > 10.233.25.222.80
ovn0                    In      2.2.98.7.42462 > 10.233.25.222.80
ovn0                    Out   100.64.0.2.37075 > 10.222.0.120.80
7fc526e9ec01_h  Out   100.64.0.2.37075 > 10.222.0.120.80

同时我通过iptables的TRACE跟踪了请求流量的iptables匹配情况，请求流量转入ovn0后会进入宿主机网络，（kube-proxy为ipvs模式）ipvs会在本地起dummy网卡（kube-ipvs0）网卡上会绑定所有的cluster ip，由于dummy网卡的存在，访问cluster ip的流量会被判定为访问本机的流量，因此进入INPUT链，在进入INPUT后被 ipvs 将cluster ip dnat为endpoints的ip，上述例子中即为10.222.0.120，在进入OUTPUT链后，根据宿主机的路由（route -n），访问10.222.0.0/18的流量会送至网关100.64.0.1，通过ovn0网卡，因此源地址被snat成了ovn0的地址（100.64.0.2.3）。

3、问题在于响应流量的路径

以下给出响应流量的Tcpdump相关路径：
7fc526e9ec01_h P      10.222.0.120.80 > 100.64.0.2.37075
ovn0                   In     10.222.0.120.80 > 100.64.0.2.37075
bond1.1807        Out  10.233.25.222.80 > 2.2.98.7.42462
b2ba4d537f3a_h Out  10.233.25.222.80 > 2.2.98.7.42462

响应报文会重新会送给ovn0然后经过宿主机的协议栈，响应报文会被还原为cluster ip和underlay ip，我这里之所以走bond1.1807，是因为宿主机上并没有2.2.98.0/24的路由，因此走了默认路由bond1.1807，而刚好bond1.1807和2.2.98.0/24是能访问通的，因此能够通。

4、响应流量的路由情况有两种
4.1 Underlay关联的Provider-network绑定的物理网卡上存在IP地址
比如配置了Provider-network为external，绑定ens33，kube-ovn-cni会在ovs添加ens33网卡后，将ens33的ip绑定到br-external的网桥上，然后配置宿主机路由，将ens33的网段的路由指向br-external。这种情况下响应报文就从br-external回来了。

4.2 Underlay关联的Provider-network绑定的物理网卡上不存在IP地址
这种情况出现在绑定bond网卡上，比如绑定bond1，然后Underlay网络还绑了VLAN，ovs给流打上VLAN后通过bond1出去。
由于bond1上没有相应的Underlay网络的ip/cidr，所以kube-ovn-cni不会配置路由，这时回程路由就只能走宿主机默认路由了，如果默认路由和你的Underlay网络能通，那就能访问成功，反之则不能成功访问。

BUG FIX
1、在Underlay绑定的物理网卡上存在IP地址的情况下，继续走br-external的路由
2、在不存在IP的情况下，添加Underlay CIDR通过100.64.0.1走ovn0的对称路由

路由添加条件：
1、子网U2O连接的是默认VPC
2、子网LB关闭且U2O开启

### Steps To Reproduce

Underlay绑定的物理网卡无IP地址或者手动删除Underlay到br-external的路由应该都能复现问题

### Current Behavior

 某些情况下开启U2O、关闭LB无法访问Service Cluster IP

### Expected Behavior

 开启U2O、关闭LB能够访问Service Cluster IP

### PR DESCRIPTION
**Pull request by zcq98**
# Pull Request

- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Bug fixes

## Which issue(s) this PR fixes

Fixes #4038

### CODE REVIEW

**Review by bobz965 — APPROVED** [review]

**Comment by bobz965** [comment]
@changluyi @zhangzujian please help review, thanks!

**Comment by zhangzujian** [comment]
Node ips may be in the subnet cidr. If we need to support u2o subnets with LB disabled, we should check node routes to ensure the route to be added will not affect current node networking.

**Comment by zcq98** [comment]
> Node ips may be in the subnet cidr. If we need to support u2o subnets with LB disabled, we should check node routes to ensure the route to be added will not affect current node networking.

如果node节点上存在underlay subnet cidr的明细路由，是不会将它删除后新增指向ovn0的路由的，只有在node上不存在目标网段为underlay subnet cidr时才会加上这条路由

File: pkg/daemon/controller_linux.go [hunk 1]
```diff
@@ -199,7 +199,10 @@ func (c *Controller) reconcileRouters(event *subnetEvent) error {
 	joinCIDR := make([]string, 0, 2)
 	cidrs := make([]string, 0, len(subnets)*2)
 	for _, subnet := range subnets {
-		if (subnet.Spec.Vlan != "" && !subnet.Spec.LogicalGateway) || subnet.Spec.Vpc != c.config.ClusterRouter || !subnet.Status.IsReady() {
+		if !subnet.Status.IsReady() ||
+			subnet.Spec.Vpc != c.config.ClusterRouter ||
+			(subnet.Spec.Vlan != "" && !subnet.Spec.LogicalGateway && !subnet.Spec.U2OInterconnection) ||
+			(subnet.Spec.Vlan != "" && !subnet.Spec.LogicalGateway && subnet.Spec.U2OInterconnection && (subnet.Spec.EnableLb != nil && *subnet.Spec.EnableLb)) {
```
Comment (zhangzujian) [inline pkg/daemon/controller_linux.go:205]:
```suggestion
			(subnet.Spec.Vlan != "" && !subnet.Spec.LogicalGateway && (!subnet.Spec.U2OInterconnection || (subnet.Spec.EnableLb != nil && *subnet.Spec.EnableLb))) {
```

**Review by zhangzujian — APPROVED** [review]

**Pull request merged**